include/tvm/runtime/container.h

```diff
@@ -188,6 +188,709 @@ class InplaceArrayBase {
+          continue;
+        }
+        ObjectPtr<ArrayNode> copy = ArrayNode::make(p->capacity_);
+        // copy [0, i)
+        int64_t& size = copy->size_ = 0;
+        for (ObjectRef* old_iter = p->MutableBegin(); size < i;) {
+          copy->EmplaceInit(size++, *old_iter++);
+        }
+        // copy [i]
+        copy->EmplaceInit(size++, std::move(new_elem));
+        // complete transformation in (i, size)
+        while (size < p->size_) {
+          T old_elem = DowncastNoCheck<T>(*iter++);
+          T new_elem = fmutate(old_elem);
+          copy->EmplaceInit(size++, std::move(new_elem));
+        }
+        data_ = std::move(copy);
+        break;
+      }
+    }
+  }
+
+  /*!
+   * \brief reset the array to content from iterator.
+   * \param begin begin of iterator
+   * \param end end of iterator
+   * \tparam IterType The type of iterator
+   */
+  template <typename IterType>
+  inline void Assign(IterType begin, IterType end) {
+    const int64_t cap = std::distance(begin, end);
+    CHECK_GE(cap, 0) << "ValueError: cannot construct an Array of negative size";
+    ObjectPtr<ArrayNode> p = ArrayNode::make(cap);
+    for (int64_t& i = p->size_ = 0; begin != end; ++i, ++begin) {
+      p->EmplaceInit(i, *begin);
+    }
+    data_ = std::move(p);
+  }
+
+  /*! \brief specify container node */
+  using ContainerType = ArrayNode;
+
+ private:
+  /*!
+   * \brief Move or copy the ArrayNode to new address with the given capacity
+   * \param capacity The capacity requirement of the new address
+   */
+  inline void SwitchContainer(const int64_t capacity) {
+    CHECK_GE(capacity, 0);
+    ObjectPtr<ArrayNode> p = ArrayNode::make(capacity);
+    if (data_ != nullptr) {
+      ArrayNode* data = GetArrayNode();
+      int64_t size = data->size_;
+      ObjectRef* from = data->MutableBegin();
+      ObjectRef* to = p->MutableBegin();
+      if (data_.unique()) {
+        for (int64_t& i = p->size_ = 0; i < size; ++i) {
+          new (to++) ObjectRef(std::move(*from++));
+        }
+      } else {
+        for (int64_t& i = p->size_ = 0; i < size; ++i) {
+          new (to++) ObjectRef(*from++);
+        }
+      }
+    }
+    data_ = std::move(p);
+  }
+
+  /*!
+   * \brief Move elements in [idx, size) backward by `numel`
+   * \param idx The position of the first element to be moved
+   * \param numel The distance for the moving
+   */
+  inline void MoveBackwards(const int64_t idx, const int64_t numel) {
+    ArrayNode* p = GetArrayNode();
+    const int64_t size = p->size_;
+    {
+      CHECK(0 <= idx && idx <= size)
+          << "ValueError: cannot insert at index " << idx << ", because Array size is " << size;
+      if (size + numel > p->capacity_) {
+        int64_t cap = p->capacity_ * ArrayNode::INC_FACTOR;
+        cap = std::max(cap, ArrayNode::INIT_SIZE);
+        cap = std::max(cap, size + numel);
+        SwitchContainer(cap);
+      }
+    }
+    p = CopyOnWrite();
+    ObjectRef* begin = p->MutableBegin();
```

Comment:
Make this part of the code be membr func of ArrayNode. Add a bit more comment about exception safety. Perhaps the safest way to do so is first do ```Enlarge(new_size, ObjectRef(nullptr))```, so size is changed to the new size, then run the move logic. This way we are safe if there is an exception during the move